docs/odcs.rst

```diff
@@ -0,0 +1,123 @@
+target (required parameter for Koji container builds). Packages in this yum repo
+are not always signed. This is not ideal as it’s trivial to produce a container
+image that should never be shipped.
+
+Another problem with the current state is that to produce an image with signed
+packages, a parameter is needed (yum repo URL). To complicate matters, this yum
+repo URL will likely be different for future builds. This poses a huge problem
+for OSBS autorebuilds. In this case, OSBS will automatically rebuild layered
+images upon detected changes to parent image. It is necessary that these
+rebuilds pick up the latest, most relevant, signed packages.
+
+The On Demand Compose Service (`odcs`_) project has been created to address
+current limitations in obtaining a yum repo with signed packages. The general
+idea is to not require the user to provide a yum repo url when building
+container images.  Instead, OSBS will use ODCS to generate a yum repo with the
+relevant signed packages based on user configuration.
+
+API Changes
+-----------
+
+The following parameters will be used by ``buildContainer`` task defined by
+``koji-containerbuild`` plugin:
+
+``odcs_signed <bool, default True>`` When set to False, OSBS will provide an
+empty list of ``sigkeys`` to ODCS. An empty list of ``sigkeys`` indicates the
+signature of packages will not be checked causing the compose to potentially
+include unsigned packages. If ODCS integration is not in place, this parameter
+has no effect.
+
+``container.yaml`` Changes
+--------------------------
+
+In git repo, alongside Dockerfile, the ``container.yaml`` file will be enhanced
+to contain a "compose" section. This section will include all the required
+information for requesting a compose to be created by ODCS::
+
+    compose:
+      # Optional, default False. Inherit compose used for building parent image
+      inherit: true
+      # Optional, default "tag".
+      type: tag
+      # Required for "tag" type. Must contain at least one item.
+      packages:
+      - package-1-name # This is a package name, not an NVR.
+      - package-2-name
+      # Optional. Default value is configured in OSBS environment
+      sigkeys:
+      - AB123456
+
+Alternatively, for modules support::
+
+    compose:
+      type: module
+      # Eventually, ODCS will support ":" separator, for now "-" is used:
+      #     https://pagure.io/odcs/issue/98
+      # Required for "module" type. Must contain at least one item.
+      modules:
+      - "module_name1:stream1"
+      - "module_name2:stream1"
+
+Behavior Changes
+----------------
+
+During an OSBS build, if ``container.yaml`` exists **and** ``compose`` key is defined:
+
+- OSBS orchestrator build will request a compose from ODCS based on specified
+  parameters in ``compose`` section. Once compose completes, build resumes and
+  the URL for the generated compose is passed to worker builds via
+  ``yum_repourls`` build parameter. *NOTE: If odcs-signed parameter was set to
```

Comment:
Passing via yum_repourls is going to break the Flatpak flow because for Flatpak, the worker builds need to be able to find out what exact module versions ODCS resolved the source specification to, and retrieve the modulemd files for those module versions. That could be worked around by passing the resolved module versions to the worker builds as well, but would be easier to just leave the current system where the compose_id is passed - the extra round-trip to ODCS to retrieve the URL should be a matter of milliseconds. 

As I mentioned above, it may also be interesting as well for higher level tooling to be able to to pass in a compose ID to be able to control the compose process exactly - at least in the modular case.

Reply:
Sold! Wasn't aware the worker builds required additional information from compose other than repo file URL.